### DISCUSSION TITLE
Custom excite method for python interface

### DISCUSSION THREAD

**Discussion by tjdphd** [thliebig/openEMS-Project#128] [General]
Am I correct in my impression that the custom excitation option available in the matlab interface is not available in the python interface? If so, is there any hope for one in the future? We're working on a project that would benefit strongly from being able to use python interface instead of matlab, but we also need to be able to have custom excitations.

**Comment by LubomirJagos** [comment]
Also today I was looking for python interface for coaxial and other ports and cannot find them, so there is more what is not there :D

I tried generate from octave .xml file with custom excitement and is defined inside <FDTD></FDTD> so you can always write your script using gaussian or other excitement definition and then at the end of your script open .xml file and modified it using python and some xml lib, here is example how custom excitement looks like:
```xml
  <FDTD NumberOfTimesteps="2000000" endCriteria="0.0001" f_max="2400000000">
    <BoundaryCond xmin="PML_8" xmax="PML_8" ymin="PML_8" ymax="PML_8" zmin="PML_8" zmax="PML_8">
    </BoundaryCond>
    <Excitation Type="10" f0="2400000000" Function="4.2*sin(2*3.14*2400000000.0*t)">
    </Excitation>
  </FDTD>
```

I bet you can also using python access CSX underlying properties and use python over CSX to set custom excitation as python interface is generated from C++ files, look here  CSXCAD.CSProperties.CSPropExcitation
    https://docs.openems.de/python/CSXCAD/CSProperties/CSPropExcitation.html#cspropexcitation

**Reply by tjdphd** [reply]
Thanks! I'll have to study this though. I'm pretty new to trying to use openEMS at all.

**Comment by thliebig** [comment]
May I ask why you need a custom excitation. I always wonder why anybody would want that since the Gaussian excitation is always the most efficient...

**Reply by tjdphd** [reply]
I can't really say more than that I need it as part of my effort. I'll be studying the underlying properties of CSX to see if I can figure out what you mean about using them as an alternative.

**Reply by tjdphd** [reply]
You mention the CSPropExcitation method, I can see vaguely how this might be related to creating a custom excitation if it also underlies other excitation types: for example, if it is used in constructing the Gaussian excitation type, then seeing how it's used for that might help me to understand how to use it for something else, but without an example use case I'm pretty lost.

**Reply by thliebig** [reply]
I did mention "CSPropExcitation"? I don't think so as it has nothing to do with a custom excite. Nor does anything in CSXCAD. This is purely an openEMS setup option.  I would need to check if you can set it using the python interface.

**Reply by LubomirJagos42** [reply]
CSPropExcitation has methods SetExcitType, SetWeightFunction, SetExcitation I think these would be enough to use and you got what you want according  to https://github.com/thliebig/CSXCAD/tree/master/python/CSXCAD since I'm on vacation I have no PC to try it, but that what I would try, and into weight function as it except arrya of dimension 3 put there functions for x,y,z and use string function which fparser is able to parse, this should be running, maybe best to experiment with python is use some good IDE like PyCharm which is capable of debugging and everything, it has community free version also, but is not expensive also for commercial development

**Reply by tjdphd** [reply]
@thliebig , sorry that was my mistake for not noticing who was responding. As I think you can see now, it was actually LubomirJagos who mentioned it to me last week. We were discussing it because I noticed that the python interface does not seem to have anything equivalent to the matlab interface's SetCustomExcite method. I'm in a situation where I really need a python interface but I also need to be able to create custom excitations.

**Reply by tjdphd** [reply]
I actually have a matlab script that I'm trying to translate into a python script using the openEMS python interface, but I'm stuck at a line which looks as follows in the matlab:

FDTD = SetCustomExcite(FDTD, f0, waveform) 

where waveform defines the precise form for the custom excitation.  I was hoping to find a python method that looks something like this e.g. 

FDTD.SetCustomExcite(f0, waveform) 

but it doesn't seem to exist in the python interface.

Thanks for any help you can offer!

**Reply by thliebig** [reply]
Yes the openEMS python interface is currently lacking this functionality as it was not a priority since it is very rarely used.

**Reply by LubomirJagos42** [reply]
No, there is no equal function but using CSPropExcitation is how to make it, best for experiments is to use PyCharm IDE and it has autocomplete function, so when you import module it should be pretty easy after you obtain excitation reference using GetExcitement().
You can try some example with custom excitation and then rewrite that one to python and compare generated .xml file if they have custom excitation done same way it's ok, at the end Matlab orpython interface are just used to generate .xml file what is input for openEMS solver.

**Reply by thliebig** [reply]
CSPropExcitation can **not** set a custom excitation!!
And NO, the python interface does not generate *.xml files like Matlab does, it generates the data using the C++ interface directly and starts openEMS directly. You can export the CSXCAD structurte into an *.xml file, but not the openEMS settings which is were a custom excitation would be defined.

**Reply by thliebig** [reply]
I can only repeat, the "SetCustomExcite" function is currently **not available in python**!

**Reply by tjdphd** [reply]
@thliebig thank you for the clarification.

**Reply by LubomirJagos42** [reply]
:D ok, but still like workaround if the xml structure is generated openEMS settings could be added as xml tags? Like it's just few tags and to handle xml file in python could not be problem

**Reply by thliebig** [reply]
I do not really see this as an option. Because the python interface cannot read/load the modified xml (with the openEMS settings). This means you would also need to start openEMS manually. Even the post-processing would not work nicely anymore. This would completely work-around the python interface. It is possible, but not really very nice... Fixing/adding the missing functions is the way to go...

**Comment by tjdphd** [comment]
@thliebig So I've been studying the project and the python interface as it currently exists, and I've tracked down the declaration of the C++  SetUpCustomExcite method of the Excitation class in the file excitation.h. At the moment I'm planning on adding something to the python interface myself that will allow me to use this with the rest of the interface, because it's absolutely crucial that my project have this functionality however rarely it may be used elsewhere. 

Related to this I have some questions that I'd appreciate your help with if you don't mind:

Assuming I manage to figure out how to make SetUpCustomExcite available via the python interface how much rebuilding is involved? Is it just a matter of running python setup.py on the python interface itself or will I need to rebuild the entire project? (I've never used cython before so I'm having to learn all that as I go). 

In case it's necessary can you **please** describe how to manually build the openEMS project under Windows 10? You make a point of not documenting a manual build under windows in your documentation, and I've spent a long time trying to figure out how to do that on my own in case I need to, but have not been successful. 

thanks for considering this request.

**Reply by tjdphd** [reply]
addendum: it's clear enough I'll have to be able to rebuild in order to do this now since doing this will mean adding some code to openems.h and openems.cpp.  So I'll just reiterate my request for help in learning how to build openEMS manually on windows.

**Reply by thliebig** [reply]
After your first request I was still a little bit hopeful that you could get at least the python part (building the extension) working on Windows. But building everything on Windows 10???  
I only do update my complete build-system on Windows every couple of years and it is always a week long project with hours of cursing and swearing. Windows is just completely broken beyond expressions... at least as far as building open source software is concerned. Or maybe I'm just too much used to Linux now after this being my only system for decades... 

**My recommendation:**  If you want to get anywhere serious in developing openEMS, install a Linux Virtual Machine (e.g. with VirtualBox) and you will just see how quick and easy it is to set everything up...

Or maybe this Linux subsystem for Windows could be an option, I don't know, I have never tried it...

**Reply by tjdphd** [reply]
I've already got that and I've already successfully built it on linux. But I need to be able to build on windows 10.

**Reply by tjdphd** [reply]
I did it on WSL

**Reply by tjdphd** [reply]
Not surprised to hear about the cursing and swearing though. Unfortunately I don't have the luxury of living entirely in the linux universe. :(

**Reply by thliebig** [reply]
If you already implemented it, tested it and so on, then just write a pull request. I will review it and potentially merge it. Then at some point I will build a new Windows Version which contains your changes?
Otherwise I still recommend to be pragmatic and not have this feature in whatever you are building? I'm doing small and large RF simulations with FDTD for many decades now and I never needed anything else than a Gaussian pulse as excitation...

**Reply by tjdphd** [reply]
Sorry for all the short responses I just keep thinking of more things to say. I actually have made progress toward the build. I'm successfully building the fparser and I think I can get as far as the configure step for CSXCAD. But the build doesn't work.

**Reply by tjdphd** [reply]
I will pass this on to my team but I don't think it's going to work for us because we'd be dependent on waiting for you to do the build whenever you decide to put yourself through the misery of it. I'm willing to do all the cursing and swearing myself if you could even just outline the process for me. I've been using what documentation you do provide here: https://docs.openems.de/install/manual.html  and looking at the error messages I get along the way to try point me in the right direction but if there's any additional advice you can provde I'd obviously really appreciate it.

**Reply by tjdphd** [reply]
As for "not having this feature in" what we're building -- that's just not an option for us. We need custom excitation one way or another.

**Reply by thliebig** [reply]
May I ask what you are working on with your team? 
Fortunately just building a new windows build is currently running without problems for me. The issue is to get everything setup and all dependencies build correctly...   Maybe I can find some time soon to give you some hints. Currently I do not have access to my Windows build machine. Please remind me if I should forget about it ;)

**Reply by tjdphd** [reply]
I don't feel at liberty to describe any specifics of the project - I'll ask the team leader what, if anything, I can tell you. The only thing I can say that's relevant to this discussion is that we are building modeling/simulation software that is meant to include allowing the user to model the excitation, and that it's crucial to our use cases to be able to do that. 

I have proposition for you though: I've actually gotten quite far in getting cmake to configure the project: basically, I've had to download, and install a bunch of dependencies and figure out how to assign values to a bunch of cmake variable so that when building CSXCAD, cmake can find all the required libraries and include directorys and cmake configuration files as the case may be, 

but I'm still getting some puzzling errors that may be pretty simple for you to resolve if you could see them. If I were to describe to you the steps I've taken to do all this would you be willing to just look over this description and tell me if there's anything I'm doing wrong or that I'm forgetting to include? This way you can help without having to work to hard or do any of the cursing and swearing :)

**Reply by Sarajevo67** [reply]
Yes, because obvious reason ;-) , Windows build is neglected, but after @biergaizi updated Qt6 support #108  MinGW build on Windows is quite straightforward.:

1. Install and update MinGW64 environment
2. Install latest dependencies (mingw64)
3. Git clone opensEMS
4. Run CMAKE 
Performances are similar as Ubuntu 22.04 build.

For MS Visual Studio build, it would help to have current VS environment (.vsproj...) at least, to avoid forging environment from scratch...

**Reply by thliebig** [reply]
MinGW is not an option if you want to build the python modules? At least that was my last information. 
But maybe @biergaizi can comment on that?
And maybe MinGW has gotten much better, but my experiences always were at least as bad as with MSVC. 

I cannot supply a vsproj as I did not go that route but instead created a nmake Makefile.
The mind-boggling part was always to get Qt+vtk compiled with MSVC on windows, because Qt does not supply prebuild MSVC compiled binaries as open-source. It does provide mingw builds for OSS, this would make everything much easier, but again, python is compiled with MSVC on Windows and you need to use the same compiler if you want to link C++ software?
But my last windows PITA is already a year or two ago, maybe with Qt6 etc. everything just got very easy?? I have my doubts...

**Reply by tjdphd** [reply]
I'm a little surprised to see you say that the bottleneck for you was Qt+vtk. Twice now I've gone through the motions - to the extent that I've been able to infer them - and I can get as far as completing the configuration step for CSXCAD successfully (The build step for CSXCAD is where I'm stuck). As for vtk and Qt I just cloned from here https://gitlab.kitware.com/vtk/vtk, built it and installed it in \~/opt/vtk and then when configuring CSXCAD added "-DVTK_DIR=~/projects/ vtk/build" to the "cmake .." command line (along with some other stuff), and cmake seemed happy with that. The problem I'm having is that when I try to build I run into problems with linking to libraries. For example the build step is expecting to find the file fparser.obj in \~/opt/openEMS/lib but it's not installed there when I install fparser. I had to go look for it and copy it to that directory. That made cmake happy when building CSXCAD but now it claims it can't open mpfr.lib even though I've set MPFR_LIBRARIES in the configure step.

**Reply by thliebig** [reply]
linking fparser.obj is already suspicious. I'm pretty sure I always linked it against a *.lib or *.dll 
As for vtk build. that "simple" build may compile and might be enough for CSXCAD, but QCSXCAD needs a vtk with built-in Qt support for example. Thus I always had to configure the vtk build quite extensively to have it build correctly and with all required options.
Again, doing all this is no magic or impossible. just when you are used to Linux and building there, it feels like pure insanity and you wonder how people dare calling Windows an "Operating System"...

**Reply by biergaizi** [reply]
> MinGW is not an option if you want to build the python modules? At least that was my last information.
But maybe @biergaizi can comment on that?

MinGW *is* an option, even for Python modules. In fact I've successfully used a development setup even more unconventional than MinGW - it's MinGW toolchain cross-compiled to GNU/Linux, so the compiler is a glibc Linux ELF executable but the output is Windown PE. Then the Windows executables were tested under Wine, this way I can test Windows-specific code without touching Windows. 

There are some problems in the Python build system, so a workaround is needed - but otherwise it works. It has been a while since I last looked about it. I don't remember which package was that, setuptools? distutils? setup.py? It has been a while since I looked about it last time. If I recall correctly, the problem is that two variants of Python exists, one is the official MSVC build, another is the unofficial MinGW build. The C extensions of both Python versions are not compatible. To make the problem worse, deep in the library of one of the main Python build systems, it assumes the use of MSVC-style modules and doesn't call the MinGW toolchain properly. 

But after I built the Cython module using compiler commands manually. It worked perfectly. 

I'm not sure if MSYS2's MinGW environment has the same problem. When I have time, I'll provide full documentation on cross-compiling for Windows.

**Reply by thliebig** [reply]
Well I would be happy if I could ditch the crazy MSVC toolchain (it's just insane!!!) for MinGW. But I don't want to force Windows user to pick a very odd and specific MinGW build Python. I want them to be able to use the "normal/default" MSVC build python on Windows and I don't think the python modules build with MinGW would be compatible then?

**Reply by thliebig** [reply]
@biergaizi Until the python interface I too found it much easier to cross-compile under Linux for Windows but it was my impression that it is impossible to make that work with default Windows python then?

**Reply by tjdphd** [reply]
@thliebig: 

> linking fparser.obj is already suspicious. I'm pretty sure I always linked it against a *.lib or *.dll
As for vtk build. that "simple" build may compile and might be enough for CSXCAD, but QCSXCAD needs a vtk with built-in Qt support for example. Thus I always had to configure the vtk build quite extensively to have it build correctly and with all required options.

Yes, I found it odd too, but that was what cmake seemed to want to do. If I don't copy fparser.obj into `~/opt/openEMS/lib`,  and attempt to run `cmake --build . --config Debug`  I get the following error message:

`LINK` : fatal error LNK1104: cannot open file 'C:\Users\tdennis\opt\openEMS\lib\fparser.obj'

Cmake error messages are all I've had to go on in trying to figure out how to build this thing. My strategy has been to run cmake, examine the errors to see what cmake wants, and then try to provide it, and try again. So when I reached this point I just copied fparser.obj to where cmake was expecting to find it and that seemed to satisfy the builder on that issue. (though of course I'm having other issues too.)

As for QCSXCAD - that wasn't even on my radar yet as I've been stuck on plain old CSXCAD. And in any case your own documentation lists it as being optional so I don't even know if I need it yet.

**Reply by Sarajevo67** [reply]
@thliebig 

> how people dare calling Windows an "Operating System"...

Well, without any intention to defend Windows as OS, this issue is more related to python.org decision to use MSVC as default CPython build compiler, for their Windows binaries distribution. Even MSVC compiled Python, requires exact MSVC version for different Python versions. That is their decision and I think we shouldn't tie our hands, by following them and tie up build for MSVC. 

One solution, as you mentioned, is MinGW Python build, which is not default, that users usually install on their Windows desktop, but it is the same for WSL or any other virtual Linux distro. It requires different Python build.

General solution, as for most of Python extensions, might be resorting to more robust C-type interface.  I think that C ABI is more stable and less prone to compiler-specific variations than the C++ ABI. This sidesteps many of the issues related to C++ name mangling, different standard libraries, and ABI incompatibilities.

Anyway, it's your (and @biergaizi ) decision. If you decide to follow up **and open this as an issue,** I would be glad to test builds on Win 10 and 11 desktops.

**Reply by biergaizi** [reply]
> General solution, as for most of Python extensions, might be resorting to more robust C-type interface. I think that C ABI is more stable and less prone to compiler-specific variations than the C++ ABI. This sidesteps many of the issues related to C++ name mangling, different standard libraries, and ABI incompatibilities.

Speaking of `ctypes`...

Why did openEMS decide to build its Python interface as a shared library anyway, instead of taking the older Matlab/Octave approach of creating a wrapper of the command-line executable? This decision was the source of the behavior mismatch between Octave and openEMS. For example, saving simulation as `.xml` is not supporting in the Python interface right now, because Octave constructs the full simulation via a CSXCAD `.xml` file, meanwhile Python does the same by calling the C++ functions in openEMS directly.

**Reply by thliebig** [reply]
Well I build the python interface like this for multiple reasons. Firstly I took it as an opportunity to learn cython and how to wrap a C++ library with python. But most importantly I was never really happy about the writing an xml file and make sure the format is correct and understands on the other sides. It always felt like a bad design. 
I like the python design much better and long term I would like to rewrite QCSXCAD/AppCSXCAD as pure python using this interface allowing to read/load/save/edit/etc. an CSXCAD structure.
Additionally I have written a simple static field solver in pure python too using this nice interface by having direct access to the important method the CSXCAD library offers.
So far I'm very happy with the python interface design. Only the Windoes build is always somewhat a pain, but luckily I do not have to update that build-system too often...
Does that make sense?

**Reply by Sarajevo67** [reply]
That sounds like you are going to fade out Matlab/Octave interface and make openEMS (which is currently almost standalone) - a Python extension?

**Reply by thliebig** [reply]
No I have no intentions to drop or fade out any interface or make openEMS less standalone. I just want to improve the CSXCAD and openEMS library API's to make it even more flexible in how these two (independent) libraries can be used...

**Reply by Sarajevo67** [reply]
Great to hear so. I hope you will find workaround for that MSVC compiler dependency, because even Python extensions (i.e. NumPy), does not require MSVC.

BTW, is it possible to avoid Qt/vtk dependencies, for openEMS build?

**Reply by thliebig** [reply]
> Great to hear so. I hope you will find workaround for that MSVC compiler dependency, because even Python extensions (i.e. NumPy), does not require MSVC.

I'm not sure what you mean by that, I'm pretty sure numpy is build on Windows with MSVC but like CSXCAD and openEMS once it is build and have e.g. a wheel you do not need anything. That said, the "normal" user does never know about the pain to setup the build-chain on Windows...

> BTW, is it possible to avoid Qt/vtk dependencies, for openEMS build?

vtk is required by openEMS, but Qt is not. That is only necessary for the GUI parts (QCSXCAD etc.)

**Reply by Sarajevo67** [reply]
@thliebig 
`I'm not sure what you mean by that, I'm pretty sure numpy is build on Windows with MSVC`
My fault, I blundered what pip3 actually did.

**Comment by thliebig** [comment]
> I don't feel at liberty to describe any specifics of the project

That sounds very ominous, please make sure that your team checks that the license you use is compatible with openEMS as it is GPL and it requires every software using it to be open source too... There might be exceptions to this depending on how you use openEMS, but it was after all the intention behind my license choice many many years ago...

**Reply by tjdphd** [reply]
We will of course honor all license requirements.  In any case, I will convey your concerns about the license to the rest of my team, and if any licensing questions arise I'll pass them along to you so we can be sure we're doing things right.

**Reply by tjdphd** [reply]
I've relayed your licensing concerns to my team, and I've got confirmation from them that we will not use openEMS in a way that violates the GPL license.

**Reply by tjdphd** [reply]
@thliebig By the way, while I'm waiting for someone on my end to help me with the Windows build (yes, we are bound and determined), I've been working on building under wsl (taking your advice to make my additions to the interface under linux and submit a pull request) and I ran into a little wrinkle while following the instructions here: https://docs.openems.de/install/manual.html . In these instructions for the openEMS build it says to use: 

`cmake ../ -DCMAKE_INSTALL_PREFIX=/home/user/openEMS -DFPARSER_ROOT_DIR=/tmp/opt -DCSXCAD_ROOT_DIR=/tmp/opt`

If I do this the build fails at the make step because cmake can't find FPARSER_ROOT_DIR. So, I changed `/tmp/opt` to `/home/tdennis/openEMS` and that fixed it. 

Point being, I'm wondering that if instruction is either outdated or a typo. Maybe in an older version you (or whoever your documentarian is) were setting `CMAKE_INSTALL_PREFIX` to `/tmp/opt` instead of `/home/user/openEMS` . If so, I just thought you might want to update it to prevent others from running into this error. Then again maybe this is wsl specific? I seem to recall you saying you hadn't tried building under wsl. Anyway best wishes.

**Comment by tjdphd** [comment]
@thliebig 

Greetings Thorsten,

So, I've spent the last little while learning Cython and I took your suggestion to add the `SetCustomExcite()` interface method under Linux first since it's easier to build manually there. 

So far, I've added the necessary code to include the `SetCustomExcite()` method, and I've tested it by using a waveform string (called `waveform`) that we've been using elsewhere via the Matlab interface, and replacing the line:

 `FDTD.SetGaussExcite(0.5*(f_start+f_stop), 0.5*(f_stop-f_start))` 

with

 `FDTD.SetCustomExcite(waveform, 0.5*(f_start+f_stop), 0.5*(f_stop-f_start))` 

and it runs just fine and produces some interesting output that is distinctly different but still sensible looking. 

You said that after I've done some testing, I could put in a pull request for you to review. I'm at the point now where I could construct more of my own tests to make sure what I'm doing is producing the expected outcome - and of course, that's what I will do barring any input from you, but I'd prefer to know what your own standards are for what would constitute an acceptably rigorous test. I want to be sure that what I've produced is up to your standards. I would also need to know how to go about making the pull request once I'm ready. I've never done that on github before. 

Meanwhile we're working on a way to automate as much as possible the process of building openEMS under windows and our PI here has expressed a desire to share our build system with you once we have it working, and I will be very happy to do just that for you when we have it up and running. 

Please let me know what you think.

Thanks, Tim
